build/Shared/NoAllocEnumerateExtensions.cs

```diff
@@ -0,0 +1,407 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+
+namespace NuGet;
+
+internal static class NoAllocEnumerateExtensions
```

Comment:
I wish this sort of thing was built in.  I also really want an analyzer to ensure we keep using this sort of thing as we add more code.